libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -0,0 +1,305 @@
+    size_t triangleCount = 0;
+
+    TangentSpaceMesh::AlgorithmHint hint;
+};
+
+struct TangentSpaceMeshOutput {
+    TangentSpaceMesh::AlgorithmHint hint;
+
+    size_t triangleCount = 0;
+    size_t vertexCount = 0;
+
+    float4* tangents = nullptr;
+    float2* uvs = nullptr;
+    float3* positions = nullptr;
+    uint3* triangles32 = nullptr;
+    ushort3* triangles16 = nullptr;
+};
+
+namespace {
+
+const uint8_t NORMALS_BIT = 0x01;
+const uint8_t TANGENTS_BIT = 0x02;
+const uint8_t UVS_BIT = 0x04;
+const uint8_t POSITIONS_BIT = 0x08;
+const uint8_t INDICES_BIT = 0x10;
+
+// Input types
+const uint8_t NORMALS = NORMALS_BIT;
+const uint8_t NORMALS_TANGENTS = NORMALS_BIT | TANGENTS_BIT;
+const uint8_t POSITIONS_INDICES = POSITIONS_BIT | INDICES_BIT;
+const uint8_t NORMALS_UVS_POSITIONS_INDICES = NORMALS_BIT | UVS_BIT | POSITIONS_BIT | INDICES_BIT;
+
+const char* HintToString(AlgorithmHint hint) {
+    switch (hint) {
+        case AlgorithmHint::DEFAULT:
+            return "DEFAULT";
+        case AlgorithmHint::MIKKTSPACE:
+            return "MIKKTSPACE";
+        case AlgorithmHint::LENGYEL:
+            return "LENGYEL";
+        case AlgorithmHint::HUGHES_MOLLER:
+            return "HUGHES_MOLLER";
+        case AlgorithmHint::FRISVAD:
+            return "FRISVAD";
+        case AlgorithmHint::FLAT_SHADING:
+            return "FLAT_SHADING";
+        case AlgorithmHint::SIGN_OF_W:
+            return "SIGN_OF_W";
+        default:
+            PANIC_POSTCONDITION("Unknown hint %u", static_cast<uint8_t>(hint));
+    }
+}
+
+AlgorithmHint selectBestDefaultAlgorithm(uint8_t inputType) {
+    AlgorithmHint outHint;
+    if (inputType & NORMALS_UVS_POSITIONS_INDICES) {
+        outHint = AlgorithmHint::MIKKTSPACE;
+    } else if (inputType & POSITIONS_INDICES) {
+        outHint = AlgorithmHint::FLAT_SHADING;
+    } else if (inputType & NORMALS_TANGENTS) {
+        outHint = AlgorithmHint::SIGN_OF_W;
+    } else {
+        ASSERT_PRECONDITION(inputType & NORMALS,
+                "Must at least have normals or (positions + indices) as input");
+        outHint = AlgorithmHint::FRISVAD;
+    }
+    return outHint;
+}
+
+AlgorithmHint selectAlgorithm(TangentSpaceMeshInput *input) {
+    uint8_t inputType = 0;
+    if (input->normals) {
+        inputType |= NORMALS_BIT;
+    }
+    if (input->tangents) {
+        inputType |= TANGENTS_BIT;
+    }
+    if (input->positions) {
+        inputType |= POSITIONS_BIT;
+    }
+    if (input->uvs) {
+        inputType |= UVS_BIT;
+    }
+    if (input->triangles32 || input->triangles16) {
+        inputType |= INDICES_BIT;
+    }
+
+    bool foundHint = false;
+    AlgorithmHint outHint = AlgorithmHint::DEFAULT;
+    switch (input->hint) {
+        case AlgorithmHint::DEFAULT:
+            outHint = selectBestDefaultAlgorithm(inputType);
+            foundHint = true;
+            break;
+        case AlgorithmHint::MIKKTSPACE:
+            if (inputType & NORMALS_UVS_POSITIONS_INDICES) {
```

Comment:
Is this meant to check if `inputType` has normals, uvs, positions, _AND_ indices? Because this will return true if any of them are set. I think you want something like:

```
if ( (inputType & NORMALS_UVS_POSITIONS_INDICES) == NORMALS_UVS_POSITIONS_INDICES) {
```

Same for the other conditions.

Reply:
Done